fem/qspace.hpp

```diff
@@ -0,0 +1,189 @@
+
+#ifndef MFEM_QSPACE
+#define MFEM_QSPACE
+
+#include "../config/config.hpp"
+#include "fespace.hpp"
+
+namespace mfem
+{
+
+/// Abstract base class for QuadratureSpace and FaceQuadratureSpace.
+/** This class represents the storage layout for QuadratureFunction%s, that may
+    be defined either on mesh elements or mesh faces. */
+class QuadratureSpaceBase
+{
+protected:
+   friend class QuadratureFunction; // Uses the offsets.
+
+   Mesh &mesh; ///< The underlying mesh.
+   int order; ///< The order of integration rule.
+   int size; ///< Total number of quadrature points.
+
+   /// @brief Entity quadrature point offset array, of size num_entities + 1.
+   ///
+   /// The quadrature point values for entity i are stored in the indices between
+   /// offsets[i] and offsets[i+1].
+   Array<int> offsets;
+   /// The quadrature rules used for each geometry type.
+   const IntegrationRule *int_rule[Geometry::NumGeom];
+
+   /// Protected constructor. Used by derived classes.
+   QuadratureSpaceBase(Mesh &mesh_, int order_ = 0)
+      : mesh(mesh_), order(order_) { }
+
+   /// Protected constructor. Used by derived classes.
+   QuadratureSpaceBase(Mesh &mesh_, Geometry::Type geom,
+                       const IntegrationRule &ir);
+
+   /// Fill the @ref int_rule array for each geometry type using @ref order.
+   void ConstructIntRules(int dim);
+
+public:
+   /// Return the total number of quadrature points.
+   int GetSize() const { return size; }
+
+   /// Return the order of the quadrature rule(s) used by all elements.
+   int GetOrder() const { return order; }
+
+   /// Return the number of entities.
+   int GetNE() const { return offsets.Size() - 1; }
+
+   /// Returns the mesh.
+   inline Mesh *GetMesh() const { return &mesh; }
+
+   /// Get the (element or face) transformation of entity @a idx.
+   virtual ElementTransformation *GetTransformation(int idx) = 0;
+
+   /// Return the geometry type of entity (element or face) @a idx.
+   virtual Geometry::Type GetGeometry(int idx) const = 0;
+
+   /// Return the IntegrationRule associated with entity @a idx.
+   const IntegrationRule &GetIntRule(int idx) const
+   { return *int_rule[GetGeometry(idx)]; }
+
+   /// @brief Returns the permuted index of the @a iq quadrature point in entity
+   /// @a idx.
+   ///
+   /// For tensor-product faces, returns the lexicographic index of the
+   /// quadrature point, oriented relative to "element 1". For QuadratureSpace%s
+   /// defined on elements (not faces), the permutation is trivial, and this
+   /// returns @a iq.
+   virtual int GetPermutedIndex(int idx, int iq) const = 0;
+
+   /// Write the QuadratureSpace to the stream @a out.
+   virtual void Save(std::ostream &out) const = 0;
+
+   virtual ~QuadratureSpaceBase() { }
+};
+
+/// Class representing the storage layout of a QuadratureFunction.
+/** Multiple QuadratureFunction%s can share the same QuadratureSpace. */
+class QuadratureSpace : public QuadratureSpaceBase
```

Comment:
Is it correct to say that I should update the mixed ceed operators to use this `QuadratureSpace` in their constructor?

Reply:
The CEED interface already supports `QuadratureFunctionCoefficient`, hopefully it should just work without any modifications. I'm not sure if this PR enables some simplifications there

Reply:
There were some very minor changes, I made them in 39c4c47